### PR TITLE
fix: update for mla collect and sol

### PR DESCRIPTION
**Pull request by YijiaZhao**
#### Overview:

fix sol compute for MLA

update collector for trtllm 1.2.0rc2

### CODE REVIEW

File: collector/trtllm/collect_mla_1_1rc2.py
Comment (tianhaox) [inline collector/trtllm/collect_mla_1_1rc2.py:1]: we use float16 to unify fp16 and bf16 for now. same for kvcache, float16 or fp8, no other precision. We can use specific bfloat16 to call into trtllm, but log the result as float16 or fp8 to keep aligned previous previous defintion.

File: src/aiconfigurator/sdk/perf_database.py
Comment (tianhaox) [inline src/aiconfigurator/sdk/perf_database.py:1]: can we fix the fmha quant mode's memory and kvcache quant mode's compute in common.py and leverage the value here

File: collector/trtllm/collect_mla_1_1rc2.py
Comment (YijiaZhao) [reply collector/trtllm/collect_mla_1_1rc2.py:1]: done

File: src/aiconfigurator/sdk/perf_database.py
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/perf_database.py:1]: already reuse FMHAQuantMode.value.compute here

File: collector/trtllm/collect_moe.py [hunk 1]
```diff
@@ -148,8 +148,8 @@ def get_moe_test_cases():
         12288,
         16384,
         20480,
-        32768,
-        65536,
+        # 32768,
```
Comment (jasonqinzhou) [inline collector/trtllm/collect_moe.py]: we don't need these data points anymore?

File: collector/collect.py [hunk 2]
```diff
@@ -422,7 +427,7 @@ def collect_trtllm(num_processes: int, ops: list[str] | None = None):
             else "trtllm.collect_moe_pre_1_0"
             if v.startswith(("0.21.0", "1.0.0"))
             else "trtllm.collect_moe"
-            if v.startswith("1.1.0")
+            if v.startswith(("1.1.0", "1.2.0"))
```
Comment (jasonqinzhou) [inline collector/collect.py:575]: Just curious that are you going to collect some data points for trtllm 1.2.x ? Will that be in this PR or in another PR?

File: src/aiconfigurator/sdk/perf_database.py [hunk 3]
```diff
@@ -2036,7 +2035,7 @@ def get_sol(
             ops = (
                 b * num_heads * 2 / 2 * (s * s * 192 + s * s * 128)
             )  # 2 for fma, 2 for causality. num_heads, for local heads
-            mem_bytes = b * num_heads * 2 * (2 * s * 192 + 2 * s * 128)  # 2 for fp16, TODO
+            mem_bytes = b * num_heads * (kvcache_quant_mode.value.memory * (s * 192 + s * 128) + 2 * (s * 192 + s * 128)) # fp16 io + fp16/fp8 kv cache, TODO fp8 io
```
Comment (jasonqinzhou) [inline src/aiconfigurator/sdk/perf_database.py]: Just curious: what is 192 and 128 in this context?

File: src/aiconfigurator/sdk/perf_database.py [hunk 4]
```diff
@@ -2071,13 +2070,17 @@ def get_sol(
...
+            if kvcache_quant_mode == common.KVCacheQuantMode.fp8:
+                quant_mode_gen = common.FMHAQuantMode.fp8
+            else:
+                quant_mode_gen = common.FMHAQuantMode.float16
             # only consider fp16 mmha
             ops = 2 * b * num_heads * 1088 * s  # 2 for fma
             # kvcache load bytes will depend on kvcache quant.
             # while input q and output might be in fp16.
-            mem_bytes = b * (num_heads * 1088 * 2 + (s - 1) * 1088 * kvcache_quant_mode.value.memory)
+            mem_bytes = b * (num_heads * 1088 * 2 + (s - 1) * 512 * kvcache_quant_mode.value.memory) # fp16 io + fp16/fp8 kv cache, TODO fp8 io
```
Comment (jasonqinzhou) [inline src/aiconfigurator/sdk/perf_database.py]: what is 1088 and 512 here?

**Comment by YijiaZhao** [comment]
h200 sanity_check results:
Gemm
image

ctx attention:
image

gen attn:
image
image

ctx mla:
image
image
image

moe:

**Comment by YijiaZhao** [comment]
moe:
WeChat09ebf2980fec929f17d528af57fd228c

File: src/aiconfigurator/sdk/perf_database.py (hunk 3 above)
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/perf_database.py]: 192 for dims of Q and K, 128 for v and output

File: src/aiconfigurator/sdk/perf_database.py (hunk 4 above)
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/perf_database.py]: q dim 576, output dim 512,  576 + 512 = 1088

File: collector/collect.py (hunk 2 above)
Comment (YijiaZhao) [reply collector/collect.py:575]: will upload on this PR for sanity check

File: collector/trtllm/collect_moe.py (hunk 1 above)
Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py]: Moe will use chunk_forward for large tokens.  besides, 32768 and 65536 is too slow to collect.

File: collector/collect.py (hunk 2 above)
Comment (tianhaox) [reply collector/collect.py:575]: 1.2 is needed to fix the mla sol problem

File: collector/collect.py [hunk 5]
```diff
@@ -222,8 +222,10 @@ def create_process_exit_error(device_id, exit_code):
             # Stall detection unchanged...
             if progress_value.value == last_progress:
                 stall_count += 1
-                if stall_count > 30:
+                if stall_count > 30 and "moe" not in func.__name__:
```
Comment (tianhaox) [inline collector/collect.py:225]: what will happen if we don't add this specific logic? Looks like sglang moe collector is very fast without tuning.

File: collector/trtllm/collect_moe.py (hunk 1 above)
Comment (tianhaox) [reply collector/trtllm/collect_moe.py]: if you remove this here, where you add the logic to collect data with num_tokens == 65536?

File: collector/trtllm/collect_moe.py
Comment (tianhaox) [inline collector/trtllm/collect_moe.py:1]: maybe we need to check when will the moe illegal mem be fixed and find a closest RC.

File: src/aiconfigurator/sdk/perf_database.py
Comment (tianhaox) [reply src/aiconfigurator/sdk/perf_database.py:1]: can you help to fix FMHAQuantMode.value.memory in this pr as well. Is that used in context attention and mla?

File: src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_mla_perf.txt
Comment (tianhaox) [inline src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_mla_perf.txt:1]: do we have fp8 compute for context mla?

File: src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_attention_perf.txt
Comment (tianhaox) [inline src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_attention_perf.txt:1]: I remember that you said in 1.2.0, there're only two options: fp8 compute+fp8kvcache, or fp16 compute+fp16 kvcache, is this for context or generation?

File: collector/trtllm/collect_moe.py
Comment (AichenF) [reply collector/trtllm/collect_moe.py:1]: maybe we can use _try_ to handle potential oom issues

File: collector/collect.py (hunk 5 above)
Comment (YijiaZhao) [reply collector/collect.py:225]: To reuse the tuned results, we packed all the tokens for same n,k in one test case. Therefore it needs more time but has fewer cases

File: collector/collect.py (hunk 2 above)
Comment (YijiaZhao) [reply collector/collect.py:575]: I've fixed sol

File: collector/trtllm/collect_moe.py (hunk 1 above)
Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py]: I added it back.

File: src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_attention_perf.txt
Comment (YijiaZhao) [reply src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_attention_perf.txt:1]: It could receive fp16 input, but cvt to fp8 by apply rope. 
It would be better to update the fp8 io for attn when we do it for gemm

File: src/aiconfigurator/sdk/perf_database.py
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/perf_database.py:1]: done

File: src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_mla_perf.txt
Comment (YijiaZhao) [reply src/aiconfigurator/systems/data/h200_sxm/trtllm/1.2.0rc1/context_mla_perf.txt:1]: no. I think it would be better to update it when we collect low-bit io gemm

File: collector/trtllm/collect_moe.py
Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py:1]: I've checked it without w4afp8 which has TMA bug

File: collector/collect_all_reduce.py
Comment (tianhaox) [inline collector/collect_all_reduce.py:1]: why we need this?

File: collector/trtllm/collect_moe.py [hunk 6]
```diff
@@ -416,66 +418,85 @@ def run_moe_torch(
...
-            )
-            logits_max_tokens = (
-                torch.randn([num_tokens_lists[max_index], num_experts]).to(router_logits_dtype).to(torch.device(device))
-            )
+    # dty run
+    torch.cuda.synchronize()
+    moe.forward(hidden_states_max_tokens, logits_max_tokens, do_finalize=not min_latency_mode)
+    torch.cuda.synchronize()
+
+    if moe_type != "w4a16_mxfp4":
```
Comment (tianhaox) [inline collector/trtllm/collect_moe.py:470]: why we need cache here, how it's reused? I'm concerned about file write and read if you allow customized path. It will be tagged as vulnerability by static risk check. Can you fix moe_tune_path to tmp, or somewhere you have write access

Comment (tianhaox) [inline collector/trtllm/collect_moe.py:470]: Is this code working with older trtllm version without w4a16_mxfp4? If not, we need a new moe collect file.

Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py:470]: Not for older versions.  W4a16_mxfp4 use triton kerne. It doesn't need autotune.

Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py:470]: Autotune process will cost lots of time. We could reuse the cache file if 
1). The new cases which has same inter_size // moe_tp and expert_num // moe_ep as anyone of the cache files.
2). If MoE collecting is not finished, next running process could reuse current cache files.

I've modified the cache file path to /tmp/moe_tune_path

File: collector/collect_all_reduce.py
Comment (YijiaZhao) [reply collector/collect_all_reduce.py:1]: I've deleted it

File: collector/trtllm/collect_moe.py (hunk 6 above)
Comment (YijiaZhao) [reply collector/trtllm/collect_moe.py:470]: I use moe_tune_path = os.path.dirname(os.path.abspath(__file__)) instead of  /tmp/moe_tune_path
